File: archive/old/abc145/src/bin/f.rs
#![allow(unused, non_snake_case, dead_code, non_upper_case_globals)]
use {
    proconio::{marker::*, *},
    std::*,
    std::{cmp::*, collections::*, convert::*, iter::*, marker::*, mem::*, ops::*},
};
macro_rules ! chmin {($ base : expr , $ ($ cmps : expr ) ,+ $ (, ) * ) => {{let cmp_min = min ! ($ ($ cmps ) ,+ ) ; if $ base > cmp_min {$ base = cmp_min ; true } else {false } } } ; }
macro_rules ! chmax {($ base : expr , $ ($ cmps : expr ) ,+ $ (, ) * ) => {{let cmp_max = max ! ($ ($ cmps ) ,+ ) ; if $ base < cmp_max {$ base = cmp_max ; true } else {false } } } ; }
macro_rules ! min {($ a : expr $ (, ) * ) => {{$ a } } ; ($ a : expr , $ b : expr $ (, ) * ) => {{std :: cmp :: min ($ a , $ b ) } } ; ($ a : expr , $ ($ rest : expr ) ,+ $ (, ) * ) => {{std :: cmp :: min ($ a , min ! ($ ($ rest ) ,+ ) ) } } ; }
macro_rules ! max {($ a : expr $ (, ) * ) => {{$ a } } ; ($ a : expr , $ b : expr $ (, ) * ) => {{std :: cmp :: max ($ a , $ b ) } } ; ($ a : expr , $ ($ rest : expr ) ,+ $ (, ) * ) => {{std :: cmp :: max ($ a , max ! ($ ($ rest ) ,+ ) ) } } ; }
/*
    ■
*/
fn main() {
    input! {//
        n:usize,k:usize,
        mut h:[i64;n],
    }
    let m = n - k;
    h.insert(0, 0);
    let mut dp = vec![vec![INF; n + 4]; n + 4];
    dp[0][0] = 0;
    for i in 0..=n {
        for j in 0..m {
            let mut now = INF;
            for k in 0..i {
                chmin!(now, dp[k][j] + max(0, h[i] - h[k]));
            }
            dp[i][j + 1] = now;
        }
    }
    let mut ans = INF;
    for i in 0..=n {
        chmin!(ans, dp[i][m]);
    }
    println!("{}", ans);
}
const INF: i64 = 1 << 61;

File: edpc/f.py
S = input()
T = input()
N = len(S)
M = len(T)

dp = [[0]*(M+1) for _ in range(N+1)]
for i in range(N):
    for j in range(M):
        if S[i] == T[j]:
            dp[i+1][j+1] = max(dp[i][j]+1, dp[i][j+1], dp[i+1][j])
        else:
            dp[i+1][j+1] = max(dp[i+1][j], dp[i][j+1])


ans = ""
i = N
j = M
while i+j > 0:
    now = dp[i][j]
    while i > 0 and now == dp[i-1][j]:
        i -= 1
    while j > 0 and now == dp[i][j-1]:
        j -= 1
    if now > dp[i-1][j-1]:
        i -= 1
        j -= 1
        ans += S[i]
        continue

print(ans[::-1])
